### DISCUSSION TITLE
Persistent File Descriptor Spikes

### DISCUSSION THREAD

**Discussion by nullmaxwell** [minio/minio#20130] [Q&A]
## NOTE
If this case is urgent, please subscribe to [Subnet](https://min.io/pricing) so that our 24/7 support team may help you faster.

In version `2024-04-06T05:26:02Z` of MinIO, we identified thousands of file descriptors attributed to each MinIO node within our small cluster of 4 nodes that is resulting in high pressure on buffer/cache within physical memory on the host machine.

Cluster info on version `2024-04-06T05:26:02Z`:
```
sh-5.1# mc admin info CLUSTER/
●  minio-node-1.█████.io:9000
   Uptime: 7 hours
   Version: 2024-04-06T05:26:02Z
   Network: 4/4 OK
   Drives: 3/3 OK
   Pool: 1

●  minio-node-2.█████.io:9000
   Uptime: 7 hours
   Version: 2024-04-06T05:26:02Z
   Network: 4/4 OK
   Drives: 3/3 OK
   Pool: 1

●  minio-node-3.█████.io:9000
   Uptime: 7 hours
   Version: 2024-04-06T05:26:02Z
   Network: 4/4 OK
   Drives: 3/3 OK
   Pool: 1

●  minio-node-4.█████.io:9000
   Uptime: 7 hours
   Version: 2024-04-06T05:26:02Z
   Network: 4/4 OK
   Drives: 3/3 OK
   Pool: 1

Pools:
   1st, Erasure sets: 1, Drives per erasure set: 12

218 GiB Used, 8 Buckets, 133,132 Objects
12 drives online, 0 drives offline
```

Example `free` output on version `2024-04-06T05:26:02Z`:
```
anon@host:~$ free 
               total        used        free      shared  buff/cache   available
Mem:        65558628    18112972      717680      123616    46727976    46592056 
Swap:       23031804         268    23031536 
```

A significant portion these file descriptors (37% of 2017 total) appeared to be associated with TCP sockets.
Excerpt of output from `lsof -l | grep minio` on version `2024-04-06T05:26:02Z` :
```
...
minio 49641 49822 minio 0 6w FIFO 0,12 0t0 2623243 pipe 
minio 49641 49822 minio 0 7u sock 0,8 0t0 2629049 protocol: TCP 
minio 49641 49822 minio 0 8u sock 0,8 0t0 2629050 protocol: TCPv6 
minio 49641 49822 minio 0 31u sock 0,8 0t0 2632823 protocol: TCPv6 
minio 49641 49822 minio 0 35u sock 0,8 0t0 2631150 protocol: TCPv6 
minio 49641 49822 minio 0 36u sock 0,8 0t0 2632900 protocol: TCP
...
```

Being aware of issue #15033 , we upgraded MinIO on each node to version `2024-07-16T23:46:41Z`, however the spikes in file descriptor counts issue still persist.

Cluster info on version `2024-07-16T23:46:41Z`:
```
sh-5.1# mc admin info CLUSTER/
●  minio-node-1.█████.io:9000
   Uptime: 4 hours
   Version: 2024-07-16T23:46:41Z
   Network: 4/4 OK
   Drives: 3/3 OK
   Pool: 1

●  minio-node-2.█████.io:9000
   Uptime: 4 hours
   Version: 2024-07-16T23:46:41Z
   Network: 4/4 OK
   Drives: 3/3 OK
   Pool: 1

●  minio-node-3.█████.io:9000
   Uptime: 4 hours
   Version: 2024-07-16T23:46:41Z
   Network: 4/4 OK
   Drives: 3/3 OK
   Pool: 1

●  minio-node-4.█████.io:9000
   Uptime: 4 hours
   Version: 2024-07-16T23:46:41Z
   Network: 4/4 OK
   Drives: 3/3 OK
   Pool: 1

Pools:
   1st, Erasure sets: 1, Drives per erasure set: 12

218 GiB Used, 8 Buckets, 133,159 Objects
12 drives online, 0 drives offline
```

`free` output on version `2024-07-16T23:46:41Z`:
```
anon@host:~$ free
               total        used        free      shared  buff/cache   available
Mem:        65558628    21800916     1244808      149132    42512904    42878492
Swap:       23031804      757184    22274620
```

After upgrading, we were able to observe a similar number of file descriptors (36% of 1868) associated with TCP sockets.

Excerpt of output from `lsof -l | grep minio` on version `2024-07-16T23:46:41Z`:
```
...
minio 50330 3139099 minio 0 5r FIFO 0,12 0t0 113857964 pipe
minio 50330 3139099 minio 0 6w FIFO 0,12 0t0 113857964 pipe
minio 50330 3139099 minio 0 7u sock 0,8 0t0 113860199 protocol: TCP
minio 50330 3139099 minio 0 8u sock 0,8 0t0 113863953 protocol: TCP
minio 50330 3139099 minio 0 9u sock 0,8 0t0 113868738 protocol: TCP
minio 50330 3139099 minio 0 10u sock 0,8 0t0 113860200 protocol: TCPv6
minio 50330 3139099 minio 0 27u sock 0,8 0t0 113863961 protocol: TCPv6
minio 50330 3139099 minio 0 31u sock 0,8 0t0 113884239 protocol: TCPv6
...
```

In an effort to capture the behavior of this issue, we recorded the total number of file descriptors used by the MinIO nodes over a ~4 hour period and were able to observe the dramatic spikes.

<img width="1071" alt="image" src="https://github.com/user-attachments/assets/4ab22e0c-acb4-42e0-bed3-fcc843b9789a">

## Expected Behavior

Predictable file descriptor resource usage should occur without dramatic spikes. Hundreds of connection sockets should not be present for a relatively low usage cluster with low API request throughput.

## Current Behavior

Thousands of file descriptors including 100s of residual (stale?) connection sockets from MinIO nodes persistently cause spikes in file descriptor usage resulting in high pressure on physical memory.

## Possible Solution

We believe that this behavior could be related to the issue #15033 

## Steps to Reproduce (for bugs)

1. Running version `2024-04-06T05:26:02Z` or `2024-07-16T23:46:41Z` for several hours/days.
2. Track file descriptor and general resource usage over time and observe.

## Context

This behavior results in very high file descriptor usage which reduces the buffer/cache on the host dramatically. 

- This cluster is primary write only and upstream jobs can only account for at most 10 concurrent object write requests issued to it every ~3-5 minutes. 
- a single object to be written to the cluster typically does not exceed 5MB. 
- 0 list requests were performed during each of our observed spikes.

## Regression
Yes.

## Your Environment

* Version used (`minio --version`):`2024-04-06T05:26:02Z` and `2024-07-16T23:46:41Z`, issue observed on both versions.
* Server setup and configuration: 
Currently 4 server nodes within a single cluster deployed on Docker.
```
sh-5.1# mc admin info CLUSTER/
●  minio-node-1.█████.io:9000
   Uptime: 4 hours
   Version: 2024-07-16T23:46:41Z
   Network: 4/4 OK
   Drives: 3/3 OK
   Pool: 1

●  minio-node-2.█████.io:9000
   Uptime: 4 hours
   Version: 2024-07-16T23:46:41Z
   Network: 4/4 OK
   Drives: 3/3 OK
   Pool: 1

●  minio-node-3.█████.io:9000
   Uptime: 4 hours
   Version: 2024-07-16T23:46:41Z
   Network: 4/4 OK
   Drives: 3/3 OK
   Pool: 1

●  minio-node-4.█████.io:9000
   Uptime: 4 hours
   Version: 2024-07-16T23:46:41Z
   Network: 4/4 OK
   Drives: 3/3 OK
   Pool: 1

Pools:
   1st, Erasure sets: 1, Drives per erasure set: 12

218 GiB Used, 8 Buckets, 133,159 Objects
12 drives online, 0 drives offline
```

* Operating System and version (`uname -a`): `Linux dhost 5.10.0-26-amd64 #1 SMP Debian 5.10.197-1 (2023-09-29) x86_64 GNU/Linux`

**Comment by klauspost** [comment]
Connections are a result of incoming requests. so if your usage is spiky FDs are spiky. We cannot go into analyzing your traffic. For that look into our [subscriptions](https://min.io/pricing).

Leaving this in discussions.

**Reply by nullmaxwell** [reply]
Sure, a substantial number of the FDs could/are related to the request throughput, however, the primary concern for this issue is the 100s of TCP sockets accounting for a significant portion of those file descriptors. Are we settling on "100s of TCP connections are necessary/normal behavior for a small cluster with little throughput" and that more than 5k FDs for a single request is typical behavior for MinIO?

**Reply by klauspost** [reply]
It is pretty normal, yes. It mostly depends on your usage.  I am pretty sure if you do tracing, you will see a more than "little throughput".

**Reply by nullmaxwell** [reply]
I am the author of the only upstream jobs that write to this MinIO cluster on a regular basis. I understand the request throughput behavior very well and can assure you that "little throughput" is an accurate representation of our usage. A single ephemeral boto3 client writing at most 3-5 objects totaling no more than 1MB at a time to a single bucket every 3 or so minutes in my opinion constitutes "little throughput" but I would be more than happy to do any additional tracing that you suggest within reason.

It would be beneficial to understand at the very least how this throughput can create such high file descriptor counts/socket counts and if it is at all related to issue #15033 that appears to be very similar.

**Reply by harshavardhana** [reply]
Capture `mc admin trace -a --stats` and see what it is..

**Reply by nullmaxwell** [reply]
Thanks for the suggestion. I ran `mc admin trace -a --stats CLUSTER` for an hour. Heres the output:
```
Duration: 1h0m4s ∙∙●
Call                     	Count         	RPM   	Avg Time	Min Time  	Max Time       	Avg TTFB	Max TTFB    	Errors	RX Avg 	TX Avg	
os.OpenFileRFd           	317074 (23.2%)	5278.7	143µs   	2.248µs   	595.296726ms   	0s      	0s          	0     	-      	-     	
scanner.ScanFolder       	311310 (22.8%)	5182.7	21.791ms	2.075274ms	14m50.82149277s	0s      	0s          	0     	-      	-     	
os.OpenFileR             	272571 (20.0%)	4537.8	106µs   	2.097µs   	75.224126ms    	0s      	0s          	0     	-      	-     	
scanner.ScanObject       	239312 (17.5%)	3984.1	1.59ms  	29.234µs  	577.454674ms   	0s      	0s          	0     	-      	-     	
os.Mkdir                 	54060 (4.0%)  	900.0 	42µs    	1.306µs   	72.388916ms    	0s      	0s          	0     	-      	-     	
os.Rename                	35172 (2.6%)  	585.5 	220µs   	1.661µs   	71.929922ms    	0s      	0s          	0     	-      	-     	
os.OpenFileW             	18535 (1.4%)  	308.6 	49µs    	13.288µs  	71.870243ms    	0s      	0s          	0     	-      	-     	
os.Remove                	16860 (1.2%)  	280.7 	28µs    	4.61µs    	37.408121ms    	0s      	0s          	0     	-      	-     	
storage.Delete           	16860 (1.2%)  	280.7 	85µs    	14.767µs  	17.012497ms    	0s      	0s          	0     	-      	-     	
storageR.DeleteFile      	12645 (0.9%)  	210.5 	458µs   	81.997µs  	17.27096ms     	458µs   	17.27096ms  	0     	139 B  	0 B   	
storage.ReadXL           	12420 (0.9%)  	206.8 	39µs    	11.775µs  	1.649256ms     	0s      	0s          	2508  	-      	-     	
storageR.ReadXL          	9315 (0.7%)   	155.1 	99µs    	44.603µs  	2.437025ms     	97µs    	2.435427ms  	0     	58 B   	402 B 	
storage.RenameData       	9012 (0.7%)   	150.0 	28.026ms	2.941004ms	235.5339ms     	0s      	0s          	0     	-      	-     	
os.RemoveAll             	7164 (0.5%)   	119.3 	115µs   	56.085µs  	77.778329ms    	0s      	0s          	0     	-      	-     	
storageR.RenameDataInline	6543 (0.5%)   	108.9 	28.934ms	3.567625ms	235.701595ms   	28.934ms	235.701595ms	0     	4.4 KiB	53 B 
```


At a glance this appears to reassure that we aren't making any listing requests to the cluster, and affirms that the highest call loads are attributed to writes (I'm assuming `OpenFileRFd`, `ScanFolder` and `OpenFileR` are related to that, please correct if wrong).

I also see that the max time on `ScanFolder` is quite high but seems to be an outlier and that there are several errors related to `stroage.ReadXL`, do we think either of these occurrences are related to the issue at hand? What else can we gather from this information to chip away at this issue?

**Reply by harshavardhana** [reply]
This is a normal behavior as such scanner establishes calls every 1m cycle - the delay there is a function of how much it is sleeping between cycles.

From what I can tell this is normal, there is no leak here it just uses connections, fds and does its job. Goes silent and does the same thing again. It's visible from the graph that you are showing.

MinIO is not silently if it's not getting used by application it is making sure your data is also safe and also does other daily routine checks.

**Reply by nullmaxwell** [reply]
Appreciate the input on this and thanks for taking the time to look.

Understanding that MinIO performs other processes internally within the cluster aside from the requests that the upstream jobs are generating, if this were a result of the request load, wouldn't we see these spikes over 3-4K FDs consistently with the request timing (2 minutes)? From the data we've gathered it appears that the FD usage spikes do not align with the request timing.

Additionally, if this so happens to be nominal behavior for MinIO, what (if any) mitigation is possible? Are there possibly configuration options to limit the behavior we're observing here?
